### PR TITLE
Fix bug in MockExecutorLoader

### PR DESCRIPTION
**Pull request by juhoautio**
Fix concurrency bug in MockExecutorLoader: it shouldn't modify the state of actual flows, so make a clone of the flows, too.

Also log job logs in `MockExecutorLoader` instead of discarding them, so that it's possible to debug failures in tests.

----

The problem was this (originally revealed when `FlowRunnerTest` still used this mock class instead of a Mockito mock):
- MockExecutorLoader kept a handle to the actual ExecutableFlow instance
- MockExecutorLoader.updateExecutableFlow set the job statuses based on the current flow status, which is not synchronized with the job runner threads
- So this wrote an old job status over the newer one in ExecutableNode

Stacktrace for that:

job4: setStatus SUCCEEDED -> QUEUED
	at azkaban.executor.ExecutableNode.setStatus(ExecutableNode.java:141)
	at azkaban.executor.ExecutableNode.applyUpdateObject(ExecutableNode.java:398)
	at azkaban.executor.ExecutableFlowBase.applyUpdateObject(ExecutableFlowBase.java:352)
	at azkaban.executor.ExecutableFlowBase.applyUpdateObject(ExecutableFlowBase.java:369)
	at azkaban.executor.MockExecutorLoader.updateExecutableFlow(MockExecutorLoader.java:120)
	at azkaban.execapp.FlowRunner.updateFlow(FlowRunner.java:322)
	at azkaban.execapp.FlowRunner.updateFlow(FlowRunner.java:316)
	at azkaban.execapp.FlowRunner.progressGraph(FlowRunner.java:552)
	at azkaban.execapp.FlowRunner.runFlow(FlowRunner.java:413)
	at azkaban.execapp.FlowRunner.doRun(FlowRunner.java:238)
	at azkaban.execapp.FlowRunner.run(FlowRunner.java:215)

Overlapping updates by 2 different threads could look something like this:

2017/05/24 12:20:54.695 +0300 INFO [JobRunner-job3-1] [ExecutableNode] job4: applyUpdateObject: RUNNING-> QUEUED
2017/05/24 12:20:55.700 +0300 INFO [JobRunner-job4-1] [JobRunner] job4: changeStatus: QUEUED -> SUCCEEDED
2017/05/24 12:20:55.701 +0300 INFO [FlowRunner-exec-1] [ExecutableNode] job4: applyUpdateObject: SUCCEEDED-> QUEUED

----

Background discussion at https://github.com/azkaban/azkaban/pull/1105.

### CODE REVIEW

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java [hunk 1]
```diff
@@ -49,7 +54,8 @@
   @Override
   public void uploadExecutableFlow(ExecutableFlow flow)
       throws ExecutorManagerException {
-    flows.put(flow.getExecutionId(), flow);
+    ExecutableFlow exFlow = ExecutableFlow.createExecutableFlowFromObject(flow.toObject());
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: Why making a copy here is not obvious without reading this change history. Could you add a comment?

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java [hunk 2]
```diff
@@ -209,21 +207,20 @@ public void testQueuedFlows() throws ExecutorManagerException, IOException {
     manager.submitExecutableFlow(flow1, testUser.getUserId());
     manager.submitExecutableFlow(flow2, testUser.getUserId());
 
-    List<ExecutableFlow> testFlows = new LinkedList<ExecutableFlow>();
-    testFlows.add(flow1);
-    testFlows.add(flow2);
+    List<Integer> testFlows = Arrays.asList(flow1.getExecutionId(), flow2.getExecutionId());
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: Why is this change necessary?

**Comment by HappyRay** [comment]
Do you think it will be better to do all the following in the future?
1. fix the node status synchronization issue
2. replace MockExecutorLoader with Mockito mock objects when possible.

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 1 above)
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: I can, but the same thing (cloning) is already done for job nodes, though..

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java (hunk 2 above)
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: Because after cloning Object.equals() can't be used.

**Comment by juhoautio** [comment]
Yes, definitely.

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: Which code path did you refer to?

How is it going to help readers understand why cloning here is needed?

File: azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java (hunk 2 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/ExecutorManagerTest.java]: I see.

**Comment by HappyRay** [comment]
If the node status field is marked volatile, will that fix the status synchronization issue? Can we then use the existing test without this change to verify the behavior? Since the node captured in the mockloader is the same instance, overwriting the node status should be fine then. Am I right?

**Comment by HappyRay** [comment]
If the node status field is marked volatile, will that fix the status synchronization issue? Can we then use the existing test without this change to verify the behavior? Since the node captured in the mockloader is the same instance, overwriting the node status should be fine then. Am I right?

**Comment by juhoautio** [comment]
IMHO it's not good that this test would modify the ExecutableNodes any way. The point is to just monitor the calls, similar to Mockito. So it's good to remove this problematic behavior in any case. Yes, volatile modifier should stabilize it, too, but I would see that as a separate issue.

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 1 above)
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: This: https://github.com/juhoautio/azkaban/blob/8272797bf60cf4009e20300e490e7838573a457c/azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java#L129

Maybe there could be a class level comment instead telling what the point of this class is (ie. capture status updates so that they can be checked in unit tests).

Should I also add comments in these places where cloning is done why it's needed?

Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: It would be great if you can add comments in both places.

My understanding is that the class is simulating a DB layer which should be a copy of the original. Thus cloning is the right choice. Am I right?

**Comment by HappyRay** [comment]
I understand now why this PR doesn't depend on the status sync fix. It's because the incorrect status is written to a copy of the original node object. Right?

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 1 above)
Comment (juhoautio) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: Yes, on point.

**Comment by juhoautio** [comment]
Correct.

**Comment by juhoautio** [comment]
@HappyRay good to go?

**Comment by juhoautio** [comment]
@HappyRay: rebased & conflicts resolved.

**Comment by HappyRay** [comment]
LGTM. I appreciate the comments you added to the code.

However mixing the style changes with the real changes makes it harder to see the real changes. I'd suggest keeping code style changes separately. This is going to be less of an issue once we complete code reformatting of the entire code base. 

Per @jamiesjc's request, we delayed reformatting some directories. This is one of them. She has some pending changes she would like to go in first.

Do you mind reverting the code style change and only keep the functional change if it is not too much trouble?

Sorry about the trouble. 

Also since the original bug as described in "The problem was this (revealed when running FlowRunnerTest that also uses this mock class): ... " has been fixed, should this change description be modified?

**Comment by juhoautio** [comment]
> Do you mind reverting the code style change and only keep the functional change

Can do, no problem.

> we delayed reformatting some directories. This is one of them

I believe you mean `ExecutorManagerTest.java`.

**Comment by juhoautio** [comment]
> Also since the original bug as described in "The problem was this (revealed when running FlowRunnerTest that also uses this mock class): ... " has been fixed, should this change description be modified?

I think the current info should be kept, I will just add that since then `FlowRunnerTest` has been changed to use `Mockito` instead of `MockExecutorLoader`.

**Comment by juhoautio** [comment]
@HappyRay now without formatting changes.

**Comment by HappyRay** [comment]
Thanks!
I mean the "executor" directory and the corresponding one in test.
